Fix scopes with selected frame

jasonLaster opened the pull request:
Fixes #579 

@JasonWeathersby the change ended up being larger than I expected and certainly larger than your one line change! The big difference is that here the scopes pane uses the selectedFrame as a data source as opposed to updating the pause info directly.

jasonLaster commented on public/js/components/Scopes.js:
```diff
@@ -3,30 +3,38 @@ const { bindActionCreators } = require("redux");
-    }))
-    .toArray();
+      contents: binding[1]
+    }));
```
immutable's api is certainly more verbose

jlongster replied:
I'd rather keep the logic of `excludeVariable` in here, it feels like over-abstracting. How come you check of `binding[1]` is truthy in the function?  You should just able tp filter out `binding[1].value.missingArguments || binding[1].value.optimizedOut;` which is pretty short.

jasonLaster replied:
that's a good point

jasonLaster commented on public/js/components/Scopes.js:
```diff
@@ -3,30 +3,38 @@ const { bindActionCreators } = require("redux");
-    .concat(bindings.get("variables").entrySeq())
-    .filter(binding => (!binding[1].hasIn(["value", "missingArguments"]) &&
-                        !binding[1].hasIn(["value", "optimizedOut"])))
+  const args = [].concat(...bindings.arguments.map(toPairs));
```
I plan on switching this to just a flatten call (probably lodash)

jlongster replied:
Shouldn't need that; the way the data is shaped you can just do `.map(x => toPairs(x)[0]` like before. It's stupid, but it's because the object only ever going to have one key and one value.

jasonLaster commented on public/js/components/Scopes.js:
```diff
@@ -96,18 +104,18 @@ function getScopes(pauseInfo) {
       }
 
-      if (vars.length) {
+      if (vars && vars.length) {
```
i'll drop this check and make sure vars is always an array

jasonLaster commented on public/js/components/Scopes.js:
```diff
@@ -116,18 +124,26 @@ const Scopes = React.createClass({
   getInitialState() {
-    return { scopes: getScopes(this.props.pauseInfo) };
+    const { pauseInfo, selectedFrame } = this.props;
+    return { scopes: getScopes(pauseInfo, selectedFrame) };
```
what was the reason we used state here? could we have changed Scopes to use `componentShouldUpdate`

jlongster replied:
It's stateful because we shouldn't keep transient objects like a tree structure in the redux state. It just is dynamically generated from the state.

JasonWeathersby commented:
Jason this is more verbose but more correct :) . My initial go was similar in that I added the selectedFrame to the scopes component and modified the getScopes function.

jasonLaster commented:
Neat. Thanks for finding it again.

What were you doing with the debugger at the time?
On Thu, Aug 25, 2016 at 12:51 PM JasonWeathersby <notifications@github.com>
wrote:

> Jason this is more verbose but more correct :) . My initial go was similar
> in that I added the selectedFrame to the scopes component and modified the
> getScopes function.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/devtools-html/debugger.html/pull/581#issuecomment-242460646>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPiYnyQVXrB8W-PC2XdL1IziQjpVnmQks5qjcf-gaJpZM4JtRHb>
> .
>

JasonWeathersby commented:
I was documenting the pause reducer and I noticed that the scopes variables where coming from pause state instead of selected frame.

jasonLaster commented:
haha - love it.

jasonLaster commented:
Interesting finding by the way, I did most of the development for this feature with this cypress test. 

```js
  it.only("(Firefox) Test Scopes", function() {
    debugPage("todomvc");
    goToSource("todo-view");
    toggleBreakpoint(33);

    // pause and check the first frame
    addTodo();
    stepIn();
  });
```

Having it meant that I could make a change and then quickly be back in the right place to debug it. Honestly, I probably could do the same thing with storybook and a fixture. I'm not sure if HMR would work with the way scope state is stored, but that could be another technique

jlongster commented:
This is great, I like that the scopes is a normal JS object as it is a bit easier to dig around in. I want to avoid converting the pause object into an immutable obj at some point too. Thanks for fixing this!

jasonLaster commented:
Just fixed a related issue where some scopes had the same sibling key because they shared the same function display name or more likely were "(anonymous)".

jlongster commented:
I don't think `Scopes` is hot-reloadable as it stands because it's a connected component. Definitely should look into how to enable it for more components.

jlongster commented:
Did you accidentally merge this? Not sure you saw my review :)

jasonLaster commented:
Nope. Missed it. Will fix that now. Thanks for checking!
